x/datadeal/keeper/reward.go

```diff
@@ -0,0 +1,58 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	distrtypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+	"github.com/medibloc/panacea-core/v2/types/assets"
+	oracletypes "github.com/medibloc/panacea-core/v2/x/oracle/types"
+)
+
+// DistributeRewards distributes reward to oracles for data verification and delivery
+func (k Keeper) DistributeRewards(ctx sdk.Context, dealID uint64, oracles map[string]*oracletypes.OracleValidatorInfo) {
+	deal, err := k.GetDeal(ctx, dealID)
+	if err != nil {
+		panic(err)
+	}
+
+	dealAddress, err := sdk.AccAddressFromBech32(deal.GetAddress())
+	if err != nil {
+		panic(err)
+	}
+
+	// calculate oracle commission
+	totalBudget := deal.GetBudget().Amount.ToDec()
+	maxNumData := sdk.NewIntFromUint64(deal.GetMaxNumData()).ToDec()
+	oracleCommission := k.oracleKeeper.GetParams(ctx).OracleCommissionRate
+	pricePerData := sdk.NewCoin(assets.MicroMedDenom, totalBudget.Quo(maxNumData).TruncateDec().Mul(oracleCommission).TruncateInt())
+
+	dealBalance := k.bankKeeper.GetBalance(ctx, dealAddress, assets.MicroMedDenom)
+	if dealBalance.IsLT(pricePerData) {
+		panic(fmt.Errorf("deal's balanace is not enough"))
+	}
+
+	totalReward := sdk.NewDecCoinsFromCoins(pricePerData)
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, dealAddress, distrtypes.ModuleName, sdk.Coins{pricePerData}); err != nil {
+		panic(err)
+	}
+
+	// calculate total voting power
+	totalVotingPower := sdk.ZeroInt()
+	for _, oracle := range oracles {
+		if oracle.IsPossibleVote() {
+			totalVotingPower = totalVotingPower.Add(oracle.BondedTokens)
+		}
+	}
+	totalVotingPowerDec := totalVotingPower.ToDec()
+
+	// distribute rewards
+	for _, oracle := range oracles {
```

Comment:
We use distributeKeeper to distribute rewards, so it doesn't seem like it would be easy to track the history.
How about adding an event here so we can check it in the block info?

Reply:
`distrKeeper.AllocateTokensToValidator()` also emits events regarding to commission and reward.
But it would be also good to add event emission of **reward distribution for data verification or delivery**.
I'll add it 👍